nlp_primitives/median_word_length.py

```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+
+
+from featuretools.primitives.base import TransformPrimitive
+from numpy import median
+from woodwork.column_schema import ColumnSchema
+from woodwork.logical_types import Double, NaturalLanguage
+
+
+class MedianWordLength(TransformPrimitive):
+    """Determines the median word length.
+
+    Description:
+        Given list of strings, determine the median
+        word length in each string. A word is defined as
+        a series of any characters not separated by a delimiter.
+        If a string is empty or `NaN`, return `NaN`.
+
+    Args:
+        delimiters_regex (str): Delimiters as a regex string for splitting text into words.
+            The default delimiters include "- [].,!?;\\n".
+
+    Examples:
+        >>> x = ['This is a test file', 'This is second line', 'third line $1,000', None]
+        >>> median_word_length = MedianWordLength()
+        >>> median_word_length(x).tolist()
+        [4.0, 4.0, 3.5, nan]
+    """
+    name = "median_word_length"
+    input_types = [ColumnSchema(logical_type=NaturalLanguage)]
+    return_type = ColumnSchema(logical_type=Double, semantic_tags={'numeric'})
+
+    default_value = 0
+
+    def __init__(self, delimiters_regex=r"[- \[\].,!\?;\n]"):
+        self.delimiters_regex = delimiters_regex
+
+    def get_function(self):
+        def get_median(words):
+            if isinstance(words, list):
+                return median([len(word) for word in words])
+
+        def median_word_length(x):
+            words = x.str.split(self.delimiters_regex)
+            return words.apply(get_median)
```

Comment:
Not sure if it would be possible or more efficient to do this in a single `apply` function call? Just wondering a bit if we are traversing the series values twice here - once to split the strings into a list of words and then a second time to get the median length value for each list. Could we do this in a single function that does both the split and the median calculation?

I'm not entirely sure which would perform better, so we might need to test out both approaches on a long series to see what is faster.

Reply:
After comparing both approaches, it seems the first is a bit faster. The underlying implementation in pandas is probably better optimized for traversing the series.

![Screen Shot 2021-12-20 at 10 04 15 AM](https://user-images.githubusercontent.com/12969559/146797950-a07dcd4b-ce7b-4e26-9a4f-f51e5a8b7e46.png)

Reply:
Ok, cool. Thanks for looking into it!